File: tests/mepylome_vs_minfi.R
library("pryr")

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)
prep <- args[1]
preps <- c("illumina", "swan", "noob")

if (!prep %in% preps) {
    cat("First command line argument must be in",
        paste(preps, collapse = ", "), "\n")
    cat("Received:", prep, "\n")
    quit(status = 1)
}


# Measure initila time and memory usage
memory0 <- mem_used()
time0 <- Sys.time()

# Load minfi
memory1 <- mem_used()
time1 <- Sys.time()
suppressMessages(suppressWarnings(library(minfi)))
time2 <- Sys.time()
memory2 <- mem_used()

time_diff <- difftime(time2, time1, units = "secs")
memory_diff <- (memory2 - memory1) / 2^20

cat("minfi\n")
cat(paste("    Import time:", time_diff, "s\n"))
cat(paste("    Memory usage:", memory_diff, "MB\n\n"))


# Load conumee2.0
memory1 <- mem_used()
time1 <- Sys.time()
suppressMessages(suppressWarnings(library(conumee2.0)))
time2 <- Sys.time()
memory2 <- mem_used()

time_diff <- difftime(time2, time1, units = "secs")
memory_diff <- (memory2 - memory1) / 2^20

cat("conumee2.0\n")
cat(paste("    Import time:", time_diff, "s\n"))
cat(paste("    Memory usage:", memory_diff, "MB\n\n"))


ensure_directory_exists <- function(dir) {
    if (!dir.exists(dir)) {
        dir.create(dir, recursive = TRUE)
    }
}


HOME_DIR <- Sys.getenv("HOME")
DEFAULT_TEST_DIR <- file.path(HOME_DIR, "Documents", "mepylome", "tests")
ensure_directory_exists(DEFAULT_TEST_DIR)

# Get all *_Grn.idat files
grn_idat_files <- list.files(
    DEFAULT_TEST_DIR,
    pattern = "_Grn.idat",
    full.names = TRUE)

# Get all IDAT basepaths
basepaths <- sub("_Grn.idat$", "", grn_idat_files)


# Read IDAT files and preprocess
time1 <- Sys.time()
for (idat_file in basepaths) {
    rgSet <- read.metharray(idat_file)
    if (prep == "illumina") {
        methyl_data <- preprocessIllumina(rgSet)
    } else if (prep == "swan") {
        methyl_data <- preprocessSWAN(rgSet)
    } else if (prep == "noob") {
        methyl_data <- preprocessNoob(rgSet)
    }
}
time2 <- Sys.time()

N <- length(basepaths)
time_diff <- difftime(time2, time1, units = "secs")
tpc <- time_diff / N
memory1 <- mem_used()


cat(paste0("Extraction time (", prep, "): ", time_diff, " s\n"))
cat(paste0("   Time per case: ", tpc, " s (No. of cases:", N, ")\n\n"))

time_diff <- difftime(time2, time0, units = "secs")
memory_diff <- (memory1 - memory0) / 2^20

cat("Total:\n")
cat(paste0("    Time (includes benchmarking utils): ", time_diff, " s\n"))
cat(paste0("    Memory usage: ", memory_diff, " MB\n"))


# Save processed data
save_data <- function(data, file_path) {
  write.csv(data, file_path, row.names = TRUE)
}

rgSet <- read.metharray(basepaths[1])
if (prep == "illumina") {
  methyl_data <- preprocessIllumina(rgSet)
} else if (prep == "swan") {
  methyl_data <- preprocessSWAN(rgSet)
} else if (prep == "noob") {
  methyl_data <- preprocessNoob(rgSet)
}

methylated <- getMeth(methyl_data)
unmethylated <- getMeth(methyl_data)

save_data(
    methylated,
    file.path(DEFAULT_TEST_DIR, paste0("minfi-", prep, "-methylated.csv")))

save_data(
    unmethylated,
    file.path(DEFAULT_TEST_DIR, paste0("minfi-", prep, "-unmethylated.csv")))

File: tests/mepylome_vs_minfi.py
import sys
import time
import pandas as pd
import numpy as np
from memory_profiler import memory_usage


prep = sys.argv[1]
preps = ["illumina", "swan", "noob"]
if prep not in preps:
    print(f"First command line argument must be in {preps}")
    print(f"Received: {prep}")
    quit()

memory0 = memory_usage()[0]
time0 = time.time()


from pathlib import Path

from mepylome import MethylData, idat_basepaths
from mepylome.utils import ensure_directory_exists

time1 = time.time()
memory1 = memory_usage()[0]
print(f"Import time: {time1 - time0} s")
print(f"Memory usage: {memory1 - memory0} MB\n")


HOME_DIR = Path.home()
DEFAULT_TEST_DIR = Path(HOME_DIR, "Documents", "mepylome", "tests")

ensure_directory_exists(DEFAULT_TEST_DIR)

idat_files = sorted(idat_basepaths(DEFAULT_TEST_DIR))


time1 = time.time()
for idat_file in idat_files:
    methyl_data = MethylData(file=idat_file, prep="illumina")
time2 = time.time()
memory1 = memory_usage()[0]

N = len(idat_files)
tpc = (time2 - time1) / N

print(f"Extraction time ({prep}): {time2 - time1} s")
print(f"  Time per case: {tpc} s (No. of cases: {N})\n")

print("Total:")
print(f"Time (includes benchmarking utils): {time2 - time0} s")
print(f"Memory usage: {memory1 - memory0} MB\n")

methyl_data = MethylData(file=idat_files[0], prep="illumina")

methyl_data.methylated.sort_index().to_csv(
    Path(DEFAULT_TEST_DIR, f"mepylome-{prep}-methylated.csv")
)
methyl_data.unmethylated.sort_index().to_csv(
    Path(DEFAULT_TEST_DIR, f"mepylome-{prep}-unmethylated.csv")
)

# Not in epic: 'cg08623843'
mepylome_path = Path(DEFAULT_TEST_DIR, f"mepylome-{prep}-methylated.csv")
minfi_path = Path(DEFAULT_TEST_DIR, f"minfi-{prep}-methylated.csv")

mepylome_df = pd.read_csv(mepylome_path, index_col=0)
minfi_df = pd.read_csv(minfi_path, index_col=0)

missing_cpgs = set(minfi_df.index) - set(mepylome_df.index)
extra_cpgs = set(mepylome_df.index) - set(minfi_df.index)

overlap_cpgs = mepylome_df.index.intersection(minfi_df.index)

len(missing_cpgs)
len(extra_cpgs)
len(overlap_cpgs)

diff = mepylome_df.loc[overlap_cpgs] - minfi_df.loc[overlap_cpgs]

diff.describe()

np.max(diff)
np.linalg.norm(diff)
np.mean(np.abs(diff))
np.std(np.abs(diff))
